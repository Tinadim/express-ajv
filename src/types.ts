import { RequestHandler, Request } from 'express';
import Ajv, { Options as AjvOptions } from 'ajv';

export interface RequestSchema {
    body?: object;
    params?: object;
    query?: object;
}

export interface ValidatorOptions {
    ajv?: Ajv.Ajv;
    ajvOptions?: AjvOptions;
    formatError?: (error: any) => any;
    getContextParams?: (req: Request) => any;
}

export type SchemaCollection = { [schemaName: string]: object };

export interface ValidatorMiddleware {
    ajv: Ajv.Ajv;
    (schemaName: string): RequestHandler;
}
