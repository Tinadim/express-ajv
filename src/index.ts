export {
    SchemaCollection,
    ValidatorMiddleware,
    ValidatorOptions,
} from './types';

export { createValidator } from './create-validator';
