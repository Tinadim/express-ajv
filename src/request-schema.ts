export function createGetRequestSchema(querySchema?: object, paramsSchema?: object) {
    const schema: { [key: string]: any } = {
        type: 'object',
        properties: {},
        required: [],
    };

    if (querySchema) {
        schema.required.push('query');
        schema.properties['query'] = querySchema;
    }

    if (paramsSchema) {
        schema.required.push('params');
        schema.properties['params'] = paramsSchema;
    }

    return schema;
}
