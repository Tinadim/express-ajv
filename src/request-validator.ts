import { forOwn, pick } from 'lodash';
import { RequestHandler, Request, Response, NextFunction } from 'express';
import Ajv, { ValidateFunction } from 'ajv';
import { SchemaCollection, ValidatorOptions } from './types';

export class RequestValidator {
    ajv: Ajv.Ajv;
    private validators: { [schemaName: string]: ValidateFunction } = {};
    private formatError: (error: any) => any;
    private getContextParams: (req: Request) => any;

    constructor(
        schemas: SchemaCollection,
        options?: ValidatorOptions,
    ) {
        this.ajv = options?.ajv ?? new Ajv(options?.ajvOptions);
        this.formatError = options?.formatError ?? this.defaultErrorFormatter;
        this.getContextParams = options?.getContextParams ?? this.defaultValidationContext;
        this.registerSchemas(schemas);
    }

    private registerSchemas(schemaCollection: { [schemaName: string]: object }) {
        forOwn(schemaCollection, (schema: object, schemaName: string) => {
            const validator = this.ajv.compile(schema);
            this.validators[schemaName] = validator;
        });
    }

    private getValidatorForSchema(schemaName: string) {
        if (!schemaName) {
            throw new Error('Please provide the schema name');
        }

        const validator = this.validators[schemaName];
        if (!validator) {
            throw new Error(`Validator for schema with name ${schemaName} not found`);
        }
        return validator;
    }

    private defaultErrorFormatter(error: any) {
        return error;
    }

    private defaultValidationContext(req: Request) {
        return {};
    }

    validateSchema(schemaName: string) {
        const validator = this.getValidatorForSchema(schemaName);

        const middleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
            const contextParams = this.getContextParams(req);
            const context = Object.assign({}, this.ajv, contextParams);
            const dataToValidate = pick(req, ['body', 'params', 'query']);

            return this.validate(dataToValidate, validator, context)
                .then(() => next())
                .catch((error) => {
                    const formattedError = this.formatError(error);
                    next(formattedError);
                });
        };

        return middleware;
    }

    private validate(data: any, validator: ValidateFunction, context: any) {
        return validator.$async
            ? this.validateAsync(data, validator, context)
            : this.validateSync(data, validator, context);
    }

    private validateSync(data: any, validator: ValidateFunction, context: any) {
        const valid = validator.call(context, data);
        return valid
            ? Promise.resolve()
            : Promise.reject(new Ajv.ValidationError(validator.errors!));
    }

    private validateAsync(data: any, validator: ValidateFunction, context: any) {
        return validator.call(context, data) as Promise<any>;
    }
}
