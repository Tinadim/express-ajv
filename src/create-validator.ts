import { isObject } from 'lodash';
import { SchemaCollection, ValidatorMiddleware, ValidatorOptions } from './types';
import { RequestValidator } from './request-validator';

export function createValidator(schemas: SchemaCollection, options?: ValidatorOptions): ValidatorMiddleware {
    if (!isObject(schemas)) {
        throw new Error('schemas should be an object');
    }

    const validator = new RequestValidator(schemas, options);
    const validatorMiddleware = (schema: string) => validator.validateSchema(schema);
    return Object.assign(validatorMiddleware, { ajv: validator.ajv }) as ValidatorMiddleware;
}
