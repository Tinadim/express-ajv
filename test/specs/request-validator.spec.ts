import Ajv from 'ajv';
import { forOwn, isFunction } from 'lodash';
import { schemas } from '../fixtures/schemas';
import { RequestValidator } from '../../src/request-validator';
import { ValidatorOptions } from '../../src/types';

describe('request-validator.spec.ts', () => {
    it('Should compile the provided schemas and generate validator functions for them', () => {
        const requestValidator = new RequestValidator(schemas) as any;
        forOwn(schemas, (schema, schemaName) => {
            const validatorFunction = requestValidator.validators[schemaName];
            expect(isFunction(validatorFunction)).toBe(true);
        });
    });

    it('Should throw an error if no schema name is provided', () => {
        const requestValidator = new RequestValidator(schemas) as any;
        expect(requestValidator.getValidatorForSchema).toThrow('Please provide the schema name');
    });

    it('Should throw an error if no validator is found for the provided schema name', () => {
        const requestValidator = new RequestValidator(schemas) as any;
        expect(() => requestValidator.getValidatorForSchema('some-random-schema-name'))
            .toThrow('Validator for schema with name some-random-schema-name not found');
    });

    describe('Request validation', () => {
        function assertErrorUndefined(error?: any) {
            expect(error).toBe(undefined);
        }

        function assertErrorMatchSnapshot(error?: any) {
            expect(error.message).toEqual('validation failed');
            const errorMessages = error.errors.map((validationError: any) => validationError.message);
            expect(errorMessages).toMatchSnapshot();
        }

        function runTest(
            request: any,
            schemaName: string,
            runAssertion: (err: any) => void,
            done: (err?: any) => void,
        ) {
            const options = {
                ajvOptions: {
                    allErrors: true,
                },
            };
            const requestValidator = new RequestValidator(schemas, options);
            const mockResponse: any = {};
            const validatorMiddleware = requestValidator.validateSchema(schemaName);
            validatorMiddleware(request, mockResponse, (error: any) => {
                runAssertion(error);
                done();
            });
        }

        function testValidRequest(schemaName: string, done: (err?: any) => void) {
            const baseData = {
                name: 'Bernardo Reis',
                age: 94,
                city: 'new-york',
            };
            const mockRequest: any = {
                query: baseData,
                body: baseData,
                params: {
                    userId: '576871e16b01de9500a46269',
                },
            };
            runTest(mockRequest, schemaName, assertErrorUndefined, done);
        }

        function testInvalidRequest(schemaName: string, done: (err?: any) => void) {
            const baseData = {
                age: -1,
                city: 'new-york',
            };
            const mockRequest: any = {
                query: baseData,
                body: baseData,
                params: {
                    userId: '1234',
                },
            };
            runTest(mockRequest, schemaName, assertErrorMatchSnapshot, done);
        }

        it('Should not throw errors for a valid get request', (done) => {
            testValidRequest('getRequest', done);
        });

        it('Should not throw errors for a valid post request', (done) => {
            testValidRequest('postRequest', done);
        });

        it('Should return an error for an invalid get request', (done) => {
            testInvalidRequest('getRequest', done);
        });

        it('Should return an error for an invalid post request', (done) => {
            testInvalidRequest('postRequest', done);
        });

        it('Should handle valid data with an asynchronous schema', (done) => {
            testValidRequest('asyncSchema', done);
        });

        it('Should handle invalid data with an asynchronous schema', (done) => {
            testInvalidRequest('asyncSchema', done);
        });
    });

    it('Should use the provided ajv instance', () => {
        const ajv = new Ajv();
        const requestValidator = new RequestValidator(schemas, { ajv });
        expect(requestValidator.ajv).toBe(ajv);
    });

    it('Should create a new ajv instance with the provided options', () => {
        const ajvOptions = {
            allErrors: true,
        };
        const requestValidator = new RequestValidator(schemas, { ajvOptions }) as any;
        const options = requestValidator.ajv._options;
        expect(options).not.toBe(ajvOptions);
    });

    it('Should use a custom formatter to format validation errors', (done) => {
        const options: ValidatorOptions = {
            errorFormatter: (error) => {
                error.message = 'my custom message';
                return error;
            },
        };
        const requestValidator = new RequestValidator(schemas, options);
        const validatorMiddleware = requestValidator.validateSchema('getRequest');
        validatorMiddleware({} as any, {} as any, (error: any) => {
            expect(error.message).toEqual('my custom message');
            done();
        });
    });

    it('Should use a custom function to get the context validation from the request', (done) => {
        const options: ValidatorOptions = {
            contextExtractor: (request: any) => {
                return { foo: request.foo };
            },
        };
        const requestValidator = new RequestValidator(schemas, options) as any;
        const mockRequest = { foo: 'bar' };
        jest.spyOn(requestValidator, 'validate').mockResolvedValue({});

        const validatorMiddleware = requestValidator.validateSchema('getRequest');
        validatorMiddleware(mockRequest as any, {} as any, () => {
            const context = requestValidator.validate.mock.calls[0][2];
            expect(context.foo).toEqual('bar');
            done();
        });
    });
});
