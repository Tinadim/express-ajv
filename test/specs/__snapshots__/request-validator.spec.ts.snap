// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request-validator.spec.ts Request validation Should handle invalid data with an asynchronous schema 1`] = `
Array [
  "should have required property 'name'",
]
`;

exports[`request-validator.spec.ts Request validation Should return an error for an invalid get request 1`] = `
Array [
  "should match pattern \\"[a-f0-9]{24}\\"",
  "should have required property 'name'",
  "should be >= 0",
]
`;

exports[`request-validator.spec.ts Request validation Should return an error for an invalid post request 1`] = `
Array [
  "should match pattern \\"[a-f0-9]{24}\\"",
  "should have required property 'name'",
  "should be >= 0",
]
`;
